dotnet/src/SemanticKernel.Abstractions/Contents/BinaryContent.cs

```diff
@@ -3,118 +3,310 @@
+    /// </summary>
+    [JsonPropertyName("data")]
+    [JsonPropertyOrder(100), JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] // Ensuring Data Uri is serialized last for better visibility of other properties.
+    public ReadOnlyMemory<byte>? Data
+    {
+        get => this.GetData();
+        set => this.SetData(value);
     }
 
     /// <summary>
-    /// Initializes a new instance of the <see cref="BinaryContent"/> class.
+    /// Indicates whether the content has binary data. If false content usually must be referenced by URI.
+    /// </summary>
+    /// <returns>True if the content has binary data, false otherwise.</returns>
+    [JsonIgnore]
+    public bool CanRead
+        => this._data is not null
+        || this._dataUri is not null;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="BinaryContent"/> class with no content.
     /// </summary>
-    /// <param name="streamProvider">The asynchronous stream provider.</param>
-    /// <param name="modelId">The model ID used to generate the content</param>
-    /// <param name="innerContent">Inner content</param>
-    /// <param name="metadata">Additional metadata</param>
     /// <remarks>
-    /// The <see cref="Stream"/> is accessed and disposed as part of either the
-    /// the <see cref="GetStreamAsync"/> or <see cref="GetContentAsync"/>
-    /// accessor methods.
+    /// Should be used only for serialization purposes.
     /// </remarks>
+    [JsonConstructor]
+    public BinaryContent()
+    {
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="BinaryContent"/> class referring to an external uri.
+    /// </summary>
+    public BinaryContent(Uri? uri = null)
+    {
+        this.Uri = uri;
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="BinaryContent"/> class for a UriData or Uri referred content.
+    /// </summary>
+    /// <param name="dataUri">The Uri of the content.</param>
     public BinaryContent(
-        Func<Task<Stream>> streamProvider,
-        string? modelId = null,
-        object? innerContent = null,
-        IReadOnlyDictionary<string, object?>? metadata = null)
-        : base(innerContent, modelId, metadata)
+        // Uri type has a ushort size limit check which inviabilizes its usage in DataUri scenarios.
+        // <see href="https://github.com/dotnet/runtime/issues/96544"/>
+        string? dataUri)
     {
-        Verify.NotNull(streamProvider, nameof(streamProvider));
+        this.DataUri = dataUri;
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="BinaryContent"/> class from a byte array.
+    /// </summary>
+    /// <param name="data">Byte array content</param>
+    /// <param name="mimeType">The mime type of the content</param>
+    public BinaryContent(
+        ReadOnlyMemory<byte> data,
+        string? mimeType)
+    {
+        Verify.NotNull(data, nameof(data));
+        if (data.IsEmpty)
```

Comment:
The previous `ImageContent` had this check, I thought it did make sense to have it the same way you are suggesting not having nullable constructors. 

Either way I can ignore it

Reply:
I'd suggest keeping the check until we have a good use case for storing null data, since it's not a breaking change to remove the check, but it is to add it back in again later if we remove it now.